### PR TITLE
Upgrade gradient chekpointing

### PR DESCRIPTION
**Pull request by yafshar**
This upgrade is necessary to resolves issues related to gradient chekpointing.
`use_reentrant` is introduced in `torch.utils.checkpoint` which should be passed explicitly. Not doing so, in higher torch version 2.4 and above will raise an exception.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by yafshar** [comment]
I added a `__MIN_TORCH_VERSION` based on `2.1.0` to do extra check and make the code compliant with older versions and avoid any break. But if we only care about `>2.1.0` some checking can be cleaned and removed

**Comment by yafshar** [comment]
I still need to add a fix for deepspeed checkpointing

**Comment by yafshar** [comment]
> I added a `__MIN_TORCH_VERSION` based on `2.1.0` to do extra check and make the code compliant with older versions and avoid any break. But if we only care about `>2.1.0` some checking can be cleaned and removed

I removed this extra checks with a moving forward approach. Raise an exception for older PyTorch < 2.1.0

**Comment by yafshar** [comment]
> I still need to add a fix for deepspeed checkpointing

The fix is provided

**Comment by yafshar** [comment]
The PR resolves issues reported in HS-3407 for `main` & `transformers_future` branches.

```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_examples.py -v -s -k test_run_clm_gpt-neox-20b_deepspeed

***** train metrics *****
  epoch                       =         2.0
  max_memory_allocated (GB)   =       94.41
  memory_allocated (GB)       =       76.74
  total_flos                  = 541209984GF
  total_memory_available (GB) =       94.62
  train_loss                  =       1.761
  train_runtime               =  0:12:29.14
  train_samples               =        2334
  train_samples_per_second    =       7.392
  train_steps_per_second      =       0.462
09/23/2024 17:43:11 - INFO - __main__ - *** Evaluate ***
[INFO|trainer.py:1822] 2024-09-23 17:43:11,026 >> Using HPU graphs for inference.
[INFO|trainer.py:1840] 2024-09-23 17:43:11,027 >>
***** Running Evaluation *****
[INFO|trainer.py:1842] 2024-09-23 17:43:11,027 >>   Num examples = 241
[INFO|trainer.py:1845] 2024-09-23 17:43:11,027 >>   Batch size = 2

***** eval metrics *****
  epoch                       =        2.0
  eval_accuracy               =     0.5369
  eval_loss                   =     2.1009
  eval_runtime                = 0:01:12.54
  eval_samples                =        241
  eval_samples_per_second     =     32.669
  eval_steps_per_second       =        2.2
  max_memory_allocated (GB)   =      94.41
  memory_allocated (GB)       =      87.95
  perplexity                  =     8.1736
  total_memory_available (GB) =      94.62
[2024-09-23 17:44:42,072] [INFO] [launch.py:351:main] Process 17427 exits successfully.
[2024-09-23 17:44:42,072] [INFO] [launch.py:351:main] Process 17424 exits successfully.
[2024-09-23 17:44:42,073] [INFO] [launch.py:351:main] Process 17425 exits successfully.
[2024-09-23 17:44:42,073] [INFO] [launch.py:351:main] Process 17429 exits successfully.
[2024-09-23 17:44:42,073] [INFO] [launch.py:351:main] Process 17430 exits successfully.
[2024-09-23 17:44:43,074] [INFO] [launch.py:351:main] Process 17426 exits successfully.
[2024-09-23 17:44:43,074] [INFO] [launch.py:351:main] Process 17428 exits successfully.
[2024-09-23 17:44:44,075] [INFO] [launch.py:351:main] Process 17431 exits successfully.
PASSED
```

```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_examples.py -v -s -k test_run_clm_gpt2-xl_deepspeed
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      31.16
  memory_allocated (GB)       =       8.55
  total_flos                  = 41067876GF
  total_memory_available (GB) =      94.62
  train_loss                  =     2.8351
  train_runtime               = 0:03:14.41
  train_samples               =       2318
  train_samples_per_second    =      97.41
  train_steps_per_second      =      0.802
09/23/2024 18:42:40 - INFO - __main__ - *** Evaluate ***
[INFO|trainer.py:1825] 2024-09-23 18:42:40,702 >> Using HPU graphs for inference.
[INFO|trainer.py:1843] 2024-09-23 18:42:40,702 >>
***** Running Evaluation *****
[INFO|trainer.py:1845] 2024-09-23 18:42:40,702 >>   Num examples = 240
[INFO|trainer.py:1848] 2024-09-23 18:42:40,703 >>   Batch size = 4
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:01<00:00,  5.23it/s]
***** eval metrics *****
  epoch                       =        2.0
  eval_accuracy               =     0.4776
  eval_loss                   =     2.5763
  eval_runtime                = 0:00:12.90
  eval_samples                =        240
  eval_samples_per_second     =    107.687
  eval_steps_per_second       =      3.739
  max_memory_allocated (GB)   =      31.16
  memory_allocated (GB)       =       8.93
  perplexity                  =    13.1484
  total_memory_available (GB) =      94.62
```

```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_examples.py -v -s -k test_sft_Qwen2-7B_multi_card

***** train metrics *****
  epoch                       =      3.7209
  max_memory_allocated (GB)   =       94.36
  memory_allocated (GB)       =       63.23
  total_flos                  = 414294080GF
  total_memory_available (GB) =       94.62
  train_loss                  =      1.6748
  train_runtime               =  0:07:19.80
  train_samples_per_second    =       7.218
  train_steps_per_second      =       0.056
100%|██████████| 3/3 [00:01<00:00,  1.82it/s]
***** eval metrics *****
  epoch                       =     3.7209
  eval_loss                   =     1.6454
  eval_runtime                = 0:00:10.68
  eval_samples                =        751
  eval_samples_per_second     =     -3.931
  eval_steps_per_second       =     -0.187
  max_memory_allocated (GB)   =      94.36
  memory_allocated (GB)       =      67.78
  perplexity                  =      5.183
  total_memory_available (GB) =      94.62
PASSED

***** train metrics *****
  epoch                       =      7.2727
  max_memory_allocated (GB)   =       94.62
  memory_allocated (GB)       =       25.25
  total_flos                  = 828883840GF
  total_memory_available (GB) =       94.62
  train_loss                  =      1.9792
  train_runtime               =  0:06:49.32
  train_samples_per_second    =     120.649
  train_steps_per_second      =       0.059
100%|██████████| 2/2 [00:00<00:00,  2.42it/s]
***** eval metrics *****
  epoch                       =     7.2727
  eval_loss                   =     1.9824
  eval_runtime                = 0:00:13.73
  eval_samples                =        751
  eval_samples_per_second     =    -70.563
  eval_steps_per_second       =     -0.218
  max_memory_allocated (GB)   =      94.62
  memory_allocated (GB)       =      34.52
  perplexity                  =     7.2601
  total_memory_available (GB) =      94.62
PASSED
```

**Comment by vidyasiv** [comment]
PR + main passes tests:
```
test_run_clm_CodeLlama-13b-Instruct-hf_deepspeed 
test_run_clm_gpt-neox-20b_deepspeed 
test_run_clm_gpt2-xl_deepspeed
```

**Review by vidyasiv — APPROVED** [review]

**Review by regisss — APPROVED** [review]
